File: src/main/java/example/websocket/client/WebSocketClientHandler.java
package example.websocket.client;

import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundMessageHandlerAdapter;
import io.netty.handler.codec.http.HttpResponse;
import io.netty.handler.codec.http.websocketx.CloseWebSocketFrame;
import io.netty.handler.codec.http.websocketx.PongWebSocketFrame;
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
import io.netty.handler.codec.http.websocketx.WebSocketClientHandshaker;
import io.netty.handler.codec.http.websocketx.WebSocketFrame;
import io.netty.util.CharsetUtil;

public class WebSocketClientHandler extends ChannelInboundMessageHandlerAdapter<Object> {

    private final WebSocketClientHandshaker handshaker;

    public WebSocketClientHandler(WebSocketClientHandshaker handshaker) {
        this.handshaker = handshaker;
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        System.out.println("WebSocket Client disconnected!");
    }

    @Override
    public void messageReceived(ChannelHandlerContext ctx, Object msg) throws Exception {
        Channel ch = ctx.channel();
        if (!handshaker.isHandshakeComplete()) {
            handshaker.finishHandshake(ch, (HttpResponse) msg);
            System.out.println("WebSocket Client connected!");
            return;
        }

        if (msg instanceof HttpResponse) {
            HttpResponse response = (HttpResponse) msg;
            throw new Exception("Unexpected HttpResponse (status=" + response.getStatus() + ", content="
                    + response.getContent().toString(CharsetUtil.UTF_8) + ")");
        }

        WebSocketFrame frame = (WebSocketFrame) msg;
        if (frame instanceof TextWebSocketFrame) {
            TextWebSocketFrame textFrame = (TextWebSocketFrame) frame;
            System.out.println("WebSocket Client received message: " + textFrame.getText());
        } else if (frame instanceof PongWebSocketFrame) {
            System.out.println("WebSocket Client received pong");
        } else if (frame instanceof CloseWebSocketFrame) {
            System.out.println("WebSocket Client received closing");
            ch.close();
        }
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        cause.printStackTrace();
        ctx.close();
    }
}

File: src/main/scala/info/crlog/higgs/old/agents/websocket/client/WebSocketClientHandler.scala
package info.crlog.higgs.agents.websocket.client

/**
 * Courtney Robinson <courtney@crlog.info>
 */

import io.netty.channel.Channel
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundMessageHandlerAdapter
import io.netty.handler.codec.http.HttpResponse
import io.netty.handler.codec.http.websocketx._
import io.netty.util.CharsetUtil
import org.msgpack.MessagePack
import collection.mutable
import io.netty.logging.{InternalLoggerFactory, InternalLogger}
import info.crlog.higgs.agents.msgpack.Interaction

class WebSocketClientHandler[T <: Interaction]
(handshaker: WebSocketClientHandshaker, listeners: mutable.Map[Class[Any], (Any) => Unit])
  extends ChannelInboundMessageHandlerAdapter[AnyRef] {
  private final val logger: InternalLogger = InternalLoggerFactory.getInstance(getClass)
  val msgpack = new MessagePack()

  override def channelInactive(ctx: ChannelHandlerContext) {
  }

  def unmarshal(data: BinaryWebSocketFrame, ctx: ChannelHandlerContext) {
    val unpacker = msgpack.createBufferUnpacker()
    unpacker.wrap(data.getBinaryData.array())
    val className = unpacker.readString()
    val clazz: Class[Any] = Class.forName(className).asInstanceOf[Class[Any]]
    msgpack.register(clazz)
    listeners.get(clazz) match {
      case None => logger.warn("Interaction of type %s received but no listeners have been registered for this type".format(className))
      case Some(clz) => {
        val msg = unpacker.read(clazz)
        msg.asInstanceOf[Interaction].context = ctx
        clz(msg)
      }
    }
  }

  def messageReceived(ctx: ChannelHandlerContext, msg: AnyRef) {
    val ch: Channel = ctx.channel
    if (!handshaker.isHandshakeComplete) {
      handshaker.finishHandshake(ch, msg.asInstanceOf[HttpResponse])
      return
    }
    if (msg.isInstanceOf[HttpResponse]) {
      val response: HttpResponse = msg.asInstanceOf[HttpResponse]
      throw new Exception("Unexpected HttpResponse (status=" + response.getStatus + ", content=" + response.getContent.toString(CharsetUtil.UTF_8) + ")")
    }
    val frame: WebSocketFrame = msg.asInstanceOf[WebSocketFrame]
    if (frame.isInstanceOf[BinaryWebSocketFrame]) {
      unmarshal(frame.asInstanceOf[BinaryWebSocketFrame], ctx)
    }
    else if (frame.isInstanceOf[PongWebSocketFrame]) {
      ctx.channel().write(new TextWebSocketFrame(frame.getBinaryData))
    }
    else if (frame.isInstanceOf[CloseWebSocketFrame]) {
      ch.close
    }
  }

  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
    cause.printStackTrace
    ctx.close
  }

}

